data_structures/trie/implementation.py

```diff
@@ -68,14 +70,25 @@ def remove(self, word: str, times: int):
             word (str): The word to remove.
             times (int): How many occurrences to remove.
         """
+        trace = ArrayBasedStack()
         cur_node = self.root
         for symbol in word:
             next_node = cur_node.get_node(symbol)
             if next_node is None:
                 return
+            trace.push([cur_node, symbol])
             cur_node = next_node
+
         if cur_node.final:
             cur_node.cnt = max(0, cur_node.cnt - times)
+            if cur_node.cnt == 0:
+                cur_node.final = False
+            if cur_node.cnt == 0 and len(cur_node.edges) == 0:
```

Comment:
I think we should move this part to a private method of the Trie class. Something like `_clean_nodes` or similar (the method name is up to you, just make it descriptive enough). Then, from the `remove` method we can call `_clean_nodes(trace)`. Agree?

The main reason is to be able to split responsibility here. The `remove` method should remove strings in the Trie, and the `_clean_nodes` method should clean unreferenced nodes.

Reply:
I agree with you on splitting responsibilities.
About your question about removing the root node, it's not removed by the algorithm, because what we remove are edges between the vertices (and Python's garbage collector handles removing the unreferenced objects).


```Python
...
parent_node.edges.pop(symbol)
...
```

Maybe just for safety on future changes, it's worth adding a test for that.